File: submissions/Chimoney-Python/chimoney/Account.py
from chimoney import BaseAPI


class Account(BaseAPI):
    """
    Account Endpoints Wrapper

    This class wraps the Account endpoints of the Chi Money API.

    list of endpoints:
        - transactions_by_issue_id
        - all_transaction
        - transaction_by_id
        - account_transfer
        - delete_unpaid_transaction
    """

    def trasactions_by_issue_id(self, issue_id, sub_account=None) -> dict:
        """
        This function returns a list of transactions by issue ID.

        Args:
            issue_id (str): The issue ID of the transaction.
            sub_account (str):  The sub account of the transaction.

        Returns:
            The JSON response fron the Chimoney API
        """
        if not isinstance(issue_id, str):
            raise TypeError("Issue ID must be a string.")

        if not issue_id:
            raise ValueError("Issue ID is required.")

        params = {"issueId": issue_id}
        payload = {}
        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "POST", "/v0.2/accounts/issue-id-transactions", params=params, data=payload
        )

    def all_transaction(self, sub_account=None) -> dict:
        """
        This function returns a list of transactions by account.

        Args:
            sub_account(str): The sub account of the transaction.

        Returns:
            The JSON response from the Chimoney API
        """
        payload = {}
        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "POST",
            "/v0.2/accounts/transactions",
            data=payload,
        )

    def transaction_by_id(self, transaction_id, sub_account=None) -> dict:
        """
        This function returns a transaction by ID.

        Args:
            transaction_id(str): The ID of the transaction.
            sub_account(str): The sub account of the transaction.

        Returns:
            The JSON response from the Chimoney API
        """
        if not isinstance(transaction_id, str):
            raise TypeError("Transaction ID must be a string.")

        if not transaction_id:
            raise ValueError("Transaction ID is required.")

        params = {"id": transaction_id}
        payload = {}
        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "GET", "/v0.2/accounts/transaction", params=params, data=payload
        )

    def account_transfer(self, receiver, amount, wallet, sub_account=None) -> dict:
        """
        This function transfers funds from one account to another.

        Args:
            reciever(str): The receiver of the funds.
            amount(float): The amount of the funds.
            wallet(str): The wallet of the funds.
            sub_aacount(str): The sub account of the transaction.

        Returns:
            The JSON response from Chimoney API
        """
        if not isinstance(receiver, str):
            raise TypeError("Receiver must be a string.")

        if not isinstance(amount, int):
            raise TypeError("Amount must be an integer.")

        if not isinstance(wallet, str):
            raise TypeError("Wallet must be a string.")

        if not receiver and amount and wallet:
            raise ValueError("Reciver, amount and wallet are required.")

        payload = {
            "receiver": receiver,
            "amount": amount,
            "wallet": wallet,
        }

        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "POST",
            "/v0.2/accounts/transfer",
            data=payload,
        )

    def delete_unpaid_transaction(self, transaction_id, sub_account=None):
        """
        This function deletes an unpaid transaction.

        Args:
            transaction_id(str): The ID of the transaction.
            sub_account(str): The sub account of the transaction

        Returns:
            The JSOn response from the Chimoney API
        """
        if not isinstance(transaction_id, str):
            raise TypeError("Transaction ID must be a string.")

        if not transaction_id:
            raise ValueError("Transaction ID is required.")

        params = {"chiRef": transaction_id}

        if sub_account:
            params["subAccount"] = sub_account
        return self._handle_request(
            "DELETE", "/v0.2/accounts/delete-unpaid-transaction", params=params
        )

File: submissions/chimoney-ts-sdk/src/modules/account/account.ts
import { HttpClient } from "../../internal/http";
import type {
  ApiResponse,
  Transaction,
  AccountTransferRequest,
  AccountTransferResponse,
} from "./types";

export interface AccountModule {
  getAllTransactions(subAccount?: string | null): Promise<ApiResponse<Transaction[]>>;
  getTransactionsByIssueID(issueID: string, subAccount?: string | null): Promise<ApiResponse<Transaction[]>>;
  getTransactionByID(transactionId: string, subAccount?: string | null): Promise<ApiResponse<Transaction>>;
  accountTransfer(req: AccountTransferRequest): Promise<ApiResponse<AccountTransferResponse>>;
  deleteUnpaidTransaction(chiRef: string, subAccount?: string | null): Promise<ApiResponse<unknown>>;
}

export function createAccountModule(http: HttpClient): AccountModule {
  return {
    getAllTransactions: (subAccount = null) =>
      http.request<ApiResponse<Transaction[]>>({
        method: "POST",
        path: "/v0.2/accounts/transactions",
        payload: subAccount ? { subAccount } : {},
      }),

    getTransactionsByIssueID: (issueID, subAccount = null) =>
      http.request<ApiResponse<Transaction[]>>({
        method: "POST",
        path: "/v0.2/accounts/issue-id-transactions",
        params: { issueID },
        payload: subAccount ? { subAccount } : {},
      }),

    getTransactionByID: (transactionId, subAccount = null) =>
      http.request<ApiResponse<Transaction>>({
        method: "POST",
        path: "/v0.2/accounts/transaction",
        params: { id: transactionId },
        payload: subAccount ? { subAccount } : {},
      }),

    accountTransfer: (req) =>
      http.request<ApiResponse<AccountTransferResponse>>({
        method: "POST",
        path: "/v0.2/accounts/transfer",
        payload: {
          receiver: req.receiver,
          amount: req.amount,
          wallet: req.wallet,
          ...(req.subAccount ? { subAccount: req.subAccount } : {}),
        },
      }),

    deleteUnpaidTransaction: (chiRef, subAccount = null) =>
      http.request<ApiResponse<unknown>>({
        method: "DELETE",
        path: "/v0.2/accounts/delete-unpaid-transaction",
        payload: {
          chiRef,
          ...(subAccount ? { subAccount } : {}),
        },
      }),
  };
}
